### PR TITLE
fix: convolutional padding argument valid and same explanation

### PR DESCRIPTION
**Pull request by samholt**
Fixed convolutional padding argument "valid" and "same" explanation

This helps other developers understand how to use the padding correctly, as currently there is no explanations for what padding "valid" or "same" mean, leading to confusion.

Please do comment on how I can get this merged,

Any feedback is most welcome

### CODE REVIEW

**Comment by samholt** [comment]
@googlebot I signed it!

File: tensorflow/python/keras/layers/convolutional.py [hunk 1]
```diff
@@ -418,6 +418,8 @@ class Conv1D(Conv):
       Specifying any stride value != 1 is incompatible with specifying
       any `dilation_rate` value != 1.
     padding: One of `"valid"`, `"causal"` or `"same"` (case-insensitive).
+      `"valid"` means no padding. `"same"` results in padding the input such
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py:421]: Let's make this more specific? `same` results in padding evenly to the left/right or up/down of the input such that output has the same height / width dimension as the input.
Same for all the following.

Comment (samholt) [reply tensorflow/python/keras/layers/convolutional.py:421]: Nice, yes will add this

Comment (samholt) [reply tensorflow/python/keras/layers/convolutional.py:421]: What about

`"same" results in padding the input such that the output has the same spatial size as the original input`

Comment (samholt) [reply tensorflow/python/keras/layers/convolutional.py]: I have made the suggested changes, making this more specific, I also corrected this in all the places that could find that it was applicable to do so. I think this will help users understand how to use padding correctly when using TensorFlow.

**Comment by samholt** [comment]
Resolved the suggested change, should be good :)

File: tensorflow/python/keras/layers/convolutional.py [hunk 2]
```diff
@@ -72,6 +72,10 @@ class Conv(Layer):
       Specifying any stride value != 1 is incompatible with specifying
       any `dilation_rate` value != 1.
     padding: One of `"valid"`,  `"same"`, or `"causal"` (case-insensitive).
+      `"valid"` means no padding. `"same"` results in padding evenly to 
+      the left/right or up/down of the input such that output has the same 
+      height/width dimension as the input. `"causal"` results in causal 
+      (dilated) convolutions, e.g. `output[t]` does not depend on `input[t+1:]`.
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py:78]: you might want to replicate the `causal` part for the rest of the PR

File: tensorflow/python/keras/layers/convolutional.py [hunk 3]
```diff
@@ -418,6 +422,9 @@ class Conv1D(Conv):
       Specifying any stride value != 1 is incompatible with specifying
       any `dilation_rate` value != 1.
     padding: One of `"valid"`, `"causal"` or `"same"` (case-insensitive).
```
Comment (kyscg) [inline tensorflow/python/keras/layers/convolutional.py:424]:
```suggestion
    padding: One of `"valid"`, `"same"` or `"causal"` (case-insensitive).
```

**Review by kyscg — CHANGES_REQUESTED** [review]
For the sake of consistency :)

Changes LGTM 👍

File: tensorflow/python/keras/layers/convolutional.py (hunk 2 above)
Comment (kyscg) [reply tensorflow/python/keras/layers/convolutional.py:78]: I'm not sure if we have `"causal"` for Conv2D and Conv3D. We should instead use the `dilation_rate` parameter.

Comment (samholt) [reply tensorflow/python/keras/layers/convolutional.py:78]: We can't get the same causal dilations with only the `dilation_rate` parameter, unless the `causal` padding variant is implemented for that layer. Therefore we can resolve this thread

**Comment by samholt** [comment]
Let me know if there is anything else I can do on this ? Otherwise should be good :)

**Comment by tanzhenyu** [comment]
> Let me know if there is anything else I can do on this ? Otherwise should be good :)

I don't think my previous comments were addressed?

**Comment by samholt** [comment]
> > Let me know if there is anything else I can do on this ? Otherwise should be good :)
> 
> I don't think my previous comments were addressed?

I did, it seems that the `Conv2D` and `Conv3D` do not support the `"casual"` padding parameter, therefore I only added the padding definitions where the `"casual"` padding parameter was already defined as an option (i.e. valid). Let me know if I am incorrect and I will update the padding options in all the associated places accordingly. However @kyscg seems to think the same as well.

**Comment by kyscg** [comment]
I agree, it is possible that we've misunderstood @tanzhenyu 's suggestions. Further clarification would we very useful.

**Comment by tanzhenyu** [comment]
> > > Let me know if there is anything else I can do on this ? Otherwise should be good :)
> > 
> > 
> > I don't think my previous comments were addressed?
> 
> I did, it seems that `Conv2D` and `Conv3D` do not support the `"casual"` padding parameter, therefore I only added the padding definitions where the `"casual"` padding parameter was already defined as an option (i.e. valid). Let me know if I am incorrect and I will update the padding options in all the associated places accordingly. However @kyscg seems to think the same as well.

That makes sense.

**Review by tanzhenyu — APPROVED** [review]
